src/model/page-config-resolver.js

```diff
@@ -283,13 +283,180 @@ class JsonLdParser {
+   * @param current the node to be verified
+   * @param root the parent to track up to
+   * @param alreadySeen used to tag already visited nodes
+   * @return {!boolean} valid node
+   * @private
+   */
+  isValidElement_(current, root, alreadySeen) {
+    for (let node = current;
+        node && !node[alreadySeen]; node = node.parentNode) {
+      node[alreadySeen] = true;
+      if (node.hasAttribute('itemscope')) {
+        const type = node.getAttribute('itemtype');
+        if (type.indexOf('http://schema.org/NewsArticle') >= 0) {
+          return true;
+        } else {
+          return false;
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Obtains the product ID that meets the requirements
+   * - child of an item of type 'NewsArticle'
+   * - Not a child of an item of type 'Section'
+   * - child of an item of type 'productID'
+   * @param {Node} root A node that is an item of type 'NewsArticle'
+   * @return {?string} product ID, if found
+   * @private
+   */
+  discoverProductId_(root) {
+    const ALREADY_SEEN = 'alreadySeenForProductInfo';
+    const nodeList = root
+        .querySelectorAll('[itemprop="productID"]');
+    for (let i = 0; nodeList[i]; i++) {
+      const element = nodeList[i];
+      const content = element.content || element.textContent;
+      const item = element.closest('[itemtype][itemscope]');
+      const type = item.getAttribute('itemtype');
+      if (type.indexOf('http://schema.org/Product') <= -1) {
+        continue;
+      }
+      if (this.isValidElement_(item.parentNode, root, ALREADY_SEEN)) {
+        return content;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Extracts microdata embedded in the DOM
+   * @return {!Microdata} microdata found in the doc
+   */
+  tryExtractConfig_() {
+    let itemId = 0;
+    const results = [];
+    const nodeList = this.doc_.getRootNode().querySelectorAll(
+        "[itemscope][itemtype='http://schema.org/NewsArticle']");
+    const domReady = this.doc_.isReady();
+    for (let i = 0; nodeList[i]; i++) {
+      const element = nodeList[i];
+      if (!domReady && !hasNextNodeInDocumentOrder(element)) {
+        continue;
+      }
+      const props = {};
+      if (this.access_ == null) {
+        const discoveredAccess = this.discoverAccess_(element);
+        if (discoveredAccess != null) {
+          this.access_ = discoveredAccess;
+          props['isAccessibleForFree'] = discoveredAccess;
+        }
+      } else {
+        props['isAccessibleForFree'] = !this.access_;
+      }
+      const discoveredProductInfo = this.discoverProductId_(element);
+      if (this.productID_ == null) {
+        if (discoveredProductInfo) {
+          this.productID_ = discoveredProductInfo;
+          props['productID'] = discoveredProductInfo;
+        }
+      } else {
+        props['productID'] = this.productID_;
+      }
+      results.push({id: ++itemId, properties: props});
+    }
+    return new Microdata(results);
```

Comment:
So, really lost why we still need this? Shouldn't it just be:

```
if (this.access_ !== null && this.productId_ !== null) {
  return new PageConfig(...);
}
```
?

Reply:
Done